### PR TITLE
Add fields for proxy credentials

### PR DESCRIPTION
**Pull request by DavidJFelix**
- proxyUser and proxyPassword must both be specified for credential proxy user.
- proxyPassword must be a securestring unless isProxySecure is manually set false
- Add Windows 7 and 8 support to metadata.json
- Change Readme to reflect Windows 7 and 8 support
- Update readme to reflect new variables
- Add erb validation tests and fix previous erb tests

### CODE REVIEW

**Comment by ffrank** [comment]
Cool, thanks so much @DavidJFelix. Sorry for being tardy in picking this up.

Are you still inclined to walk through the review process on this one? I realize it's been quite a while.

**Comment by DavidJFelix** [comment]
Sure. I'll respond if prompted. What is there left to the reviewing process?

File: README.md [hunk 1]
```diff
@@ -73,6 +73,34 @@ The optional name of the file to download onto the system.
...
 The optional http proxy address to use when downloading the file
 
+#####`proxyUser`
+The optional http proxy user to use when downloading the file. `proxyAddress` and `proxyPassword`
+must be specified or this has no effect.
+
+#####`proxyPassword`
+The optional http proxy password to use when downloading the file. `proxyAddress` and `proxyUser`
+must be specified or this has no effect. By default this value accepts secure strings. A secure
+string is (unfortunately) tied to the machine that it is ued for. To generate a secure string for
```
Comment (ffrank) [inline README.md]: "used"

File: manifests/init.pp [hunk 2]
```diff
@@ -46,7 +46,10 @@
   $url,
   $destination_directory,
   $destination_file = '',
-  $proxyAddress=''
+  $proxyAddress='',
```
Comment (ffrank) [inline manifests/init.pp]: So `$proxyAddress` was camel case all along, but I would prefer underscore downcase names, just for the reason that they are more in keeping with Puppet practices. Thoughts?

File: metadata.json [hunk 3]
```diff
@@ -11,7 +11,7 @@
   "operatingsystem_support": [
     {
     "operatingsystem":"Windows",
-    "operatingsystemrelease":[ "2008R2", "2012", "2012R2" ]
+    "operatingsystemrelease":[ "2008R2", "2012", "2012R2", "7", "8" ]
```
Comment (ffrank) [inline metadata.json]: Thanks for testing, all I wonder is

 * should this metadata change migrate into its own changeset?
 * do we have special requirements for rigorous testing before making this claim?

As for the latter, I suppose not. If you've seen it working on those platforms - great. If there are issues that were missed, those will come up eventually.

File: manifests/init.pp (hunk 2 above)
Comment (DavidJFelix) [inline manifests/init.pp]: I was just following the camel case lead. I prefer snake case also.

File: manifests/init.pp [hunk 4]
```diff
@@ -46,7 +46,10 @@
...
+  $proxyUser='',
+  $proxyPassword='',
+  $isPasswordSecure=true
```
Comment (ffrank) [inline manifests/init.pp]: I see how security seems to be a good default, but the implementation here is a bit problematic for Puppet, I feel.

Normally, you want to define node data prior to setting up a machine in the first place, and enable Puppet to make sure everything works coherently. This is inherently incompatible with this mode of password security. As such, I'm inclined to believe that `false` might be a better default here. Users that feel that the proxy credentials must be protected as vehemently as possible would alter this setting. Thoughts?

**Comment by ffrank** [comment]
I think we're down to the nitpicking stage. The change looks really good. I'm especially fond of the test coverage. Let's work out all the details and then I believe this will be good to merge quite soon.

I will need feedback from someone who can actually run this on Windows and has more knowledge, @stack72 ideally.

**Comment by ffrank** [comment]
Ah, but it *will* need to be rebased, apparently.

File: manifests/init.pp (hunk 4 above)
Comment (DavidJFelix) [inline manifests/init.pp]: I agree from a usability stance that false would be better. One thing to note here is that I believe the secure string can only be created on the node itself... so not sure how you would get the secure string for an arbitrary node. For me, I was doing masterless, so it made sense. For most it probably won't

File: metadata.json (hunk 3 above)
Comment (DavidJFelix) [inline metadata.json]: Probably could pull it out into it's own change. 7 and 8 should have powershell parity with the mentioned server versions. You really can't test that powershell is setup correctly, regardless of version, so it's really an all or none question in my mind, and I don't think none is reasonable. I did work on 7 and 8 to get this working and then used it on 2012.

File: manifests/init.pp (hunk 4 above)
Comment (ffrank) [inline manifests/init.pp]: Yes, exactly - node data cannot be pre-populated. So, many if not most users will likely forego this feature.

**Comment by ffrank** [comment]
Great, so I'd be greatful if you could

 * rebase to master
 * switch to snake-case parameters
  * the camel case `proxyAddress` should be kept but deprecated
 * make `secure` default to `false`
 * move the metadata change to a new PR
 * fix the minor documentation typo

Cheers!

**Comment by daenney** [comment]
> * make secure default to false

This sentence makes me itch, a lot. Why do we want to default `secure` to `false`? That seems to be a wrong default to have. Even if it's the practical one, degrading security should be a conscious decision made by the end user.

**Comment by DavidJFelix** [comment]
I've abandoned this pull request.

**Comment by ffrank** [comment]
@DavidJFelix sad to hear it - would you mind if we salvage it on our own?

**Comment by DavidJFelix** [comment]
@ffrank go for it. I'm not totally apathetic to the matter, so I'd be glad to help in conversation, but I've abandoned this PR for a few reasons:

1. I no longer use puppet
2. I don't think using powershell + templates is the right way to accomplish this and I think the struggle here is illustrating "doing it wrong". As I grew to know puppet, this felt wrong.
3. I didn't feel like anyone was going to use any of the code
4. Inclusion of the changes has stagnated to the point where I have to keep refreshing myself to remember where it's at.


So here goes a mind-dump of my thoughts on this:

The idea behind ruby is to create a platform agnostic scripting language right? So is it so hard for puppet manifests to also be platform agnostic? [Ruby has net/http and can issue a proxied get request, can't it?](http://stackoverflow.com/questions/4581075/how-make-a-http-get-request-using-ruby-on-rails) so why not create a puppet class that gets a file? The puppet DSL obfuscates how to call native ruby a bit, but it's still documented and possible. I think this is the way that files should be retrieved. Further, the class created should have mechanisms which allow it to operate statefully. I'm not a huge fan of this, but it's the Puppet way to recognize when the up-to-date file exists and when it should be downloaded vs when that step should be skipped.

Utilizing ruby would mean that the curl/wget packages could be deprecated and that this module would support all OS, not just windows. It also removes weird questions like how windows secures passwords, etc, like we're currently having.

**Comment by DavidJFelix** [comment]
Also, my use case is mostly masterless. Masterless is weird in that your agent is technically also the master. It may be beneficial to design the module in a way such that the download is run on the master and the file is saved to the `puppet:///` served directory and the class command simply wires an agent to this master's downloaded and served file.

**Comment by daenney** [comment]
> Masterless is weird in that your agent is technically also the master. It may be beneficial to design the module in a way such that the download is run on the master and the file is saved to the puppet:/// served directory and the class command simply wires an agent to this master's downloaded and served file.

That would not work out that great, especially with people continuously deploying their environments with r10k as the target directory on the master would get wiped. Additionally you'd be adding things to a tree of code that is normally under VCS out of band nuking any kind of audibility features and transferring the file twice, once fetching it to the master and then moving it over to the agent.

**Comment by DavidJFelix** [comment]
Probably. It felt weird suggesting that anything get dyamically pulled into `puppet:///` but the real point above is that having powershell do it isn't the right solution, in my mind.

**Comment by ffrank** [comment]
Hmpf, when flooded with github notifications I start tuning out the bigger picture of individual PRs. Such as this one. Oh well. Back up to speed now.

As it happens, some slow shmuck sat on his Puppet patches for `file { source => "http://..." }` support for quite a while, but it finally [got merged](https://github.com/puppetlabs/puppet/pull/3748#event-495045907) just recently. So Puppet `4.3` (I think) will support that. And yes, this module may grow obsolete in time. I agree on the notion that the language should allow this. (And it's in core - yay)

All that being said, it's nice that this module gives this capability to Puppet `3.x` users (as do some others), so I'm still willing to maintain it. Migrating it to types/providers, on the other hand, feels like overkill in light of of the above.

With your blessing, I'll base a new PR on your commits "soon". Thanks for all the input!

**Comment by DavidJFelix** [comment]
blessed

**Pull request closed without merging**